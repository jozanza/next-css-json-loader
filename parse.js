const css = require('css');
const transformers = require('./transformers');
const _ = require('./utils');

const joinRules = ctx => (styles, rule) => {
  const type = _.camelize(rule.type);
  if (!transformers[type])
    throw new Error(`Could not transform CSS rule: ${JSON.stringify(rule, null, 2)}`);
  return transformers[type](ctx)(styles, rule);
};

const transformRule = ctx => x => [
  _.camelize(x.type),
  joinRules(ctx)(x.type === 'keyframes' ? [] : {}, x)
];

const mergeStyles = (acc, val) => {
  const type = val[0];
  const rule = val[1];
  if (type === 'keyframes') // dedupe animations
    acc.keyframes = acc.keyframes.filter(x => x[0] !== rule[0]);
  if (!acc[type]) console.log(type)
  acc[type].push(rule);
  return acc;
};

module.exports = ctx => text => css
  .parse(text)
  .stylesheet
  .rules
  .map(transformRule(ctx))
  .reduce(mergeStyles, {
    charset: [],
    fontFace: [],
    rule: [],
    media: [],
    keyframes: []
  });
