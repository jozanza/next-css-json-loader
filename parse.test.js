import test from 'ava';
import parse from './parse';

test('basic rules', t => {
  const ctx = '';
  const input = `
  .foo { width: 100%; }`;
  const output = parse(ctx)(input);
  t.snapshot(output);
});

test('@keyframes', t => {
  const ctx = '';
  const input = `
  @keyframes bounce {
    0%, 50% { blah: .5; }
    to   { lol: 100vh; }
  }`;
  const output = parse(ctx)(input);
  t.snapshot(output);
});

test('@media', t => {
  const ctx = '';
  const input = `
  @media(min-width: 100px; max-width: 1000px) {
    .foo {
      bar: 88px;
    }
  }`;
  const output = parse(ctx)(input);
  t.snapshot(output);
});

test('@font-face', t => {
  const ctx = '';
  const input = `
  @font-face {
    font-family: 'Open Sans';
    font-style: normal;
    font-weight: 400;
    src: local('Open Sans'),
         local('OpenSans'),
         url('https://fonts.gstatic.com/s/...');
  }`;
  const output = parse(ctx)(input);
  t.snapshot(output);
});


