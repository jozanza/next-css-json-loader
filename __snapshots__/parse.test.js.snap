exports[`@font-face 1`] = `"{\"charset\":[],\"fontFace\":[{\"fontFamily\":\"\'Open Sans\'\",\"fontStyle\":\"normal\",\"fontWeight\":\"400\",\"src\":\"local(\'Open Sans\'),\\n         local(\'OpenSans\'),\\n         url(\\\"https://fonts.gstatic.com/s/...\\\")\"}],\"rule\":[],\"media\":[],\"keyframes\":[]}"`;

exports[`@keyframes 1`] = `"{\"charset\":[],\"fontFace\":[],\"rule\":[],\"media\":[],\"keyframes\":[[\"bounce\",{\"0%,50%\":{\"blah\":\".5\"},\"to\":{\"lol\":\"100vh\"}}]]}"`;

exports[`@media 1`] = `"{\"charset\":[],\"fontFace\":[],\"rule\":[],\"media\":[{\"@media(min-width: 100px; max-width: 1000px)\":{\".foo\":{\"bar\":\"88px\"}}}],\"keyframes\":[]}"`;

exports[`basic rules 1`] = `"{\"charset\":[],\"fontFace\":[],\"rule\":[{\".foo\":{\"width\":\"100%\"}}],\"media\":[],\"keyframes\":[]}"`;
